Add migration docs for 02-client refactor (#1287)
cosmos/ibc-go: docs/migrations/v3-to-v4.md

## Chains

### IS04 - Channel

The `WriteAcknowledgement` API now takes the `exported.Acknowledgement` type instead of passing in the acknowledgement byte array directly.
This is an API breaking change and as such IBC application developers will have to update any calls to `WriteAcknowledgement`.

## IBC Light Clients

> ### ClientState interface changes

Reviewer: ```suggestion
### `ClientState` interface changes
```

> The `VerifyUpgradeAndUpdateState` function has been modified. The client state and consensus state return value has been removed.

Reviewer: ```suggestion
The `VerifyUpgradeAndUpdateState` function has been modified. The client state and consensus state return values has been removed.
```

Author: Yeah, but I think it should be like this:
"The `VerifyUpgradeAndUpdateState` function has been modified. The returned client state and consensus state values has been removed."

Light clients **must** handle all management of client and consensus states including the setting of updated client state and consensus state in the client store.

The `CheckHeaderAndUpdateState` function has been split into 4 new functions:

> - `VerifyClientMessage` : verify a `ClientMessage`. A ClientMessage could be a Header, Misbehaviour, or batch update. Calls to `CheckForMisbehaviour`, `UpdateState`, and `UpdateStateOnMisbehaviour` will assume that the content of the `ClientMessage` has been verified and can be trusted. An error should be returned if the `ClientMessage` fails to verify.

Reviewer: ```suggestion
- `VerifyClientMessage` verifies a `ClientMessage`. A `ClientMessage` could be a `Header`, `Misbehaviour`, or batch update. Calls to `CheckForMisbehaviour`, `UpdateState`, and `UpdateStateOnMisbehaviour` will assume that the content of the `ClientMessage` has been verified and can be trusted. An error should be returned if the `ClientMessage` fails to verify.
```

> - `CheckForMisbehaviour` : checks for evidence of a misbehaviour in Header or Misbehaviour type.

Reviewer: ```suggestion
- `CheckForMisbehaviour` checks for evidence of a misbehaviour in `Header` or `Misbehaviour` types.
```

> - `UpdateStateOnMisbehaviour` : perform appropriate state changes on a `ClientState` given that misbehaviour has been detected and verified.

Reviewer: ```suggestion
- `UpdateStateOnMisbehaviour` performs appropriate state changes on a `ClientState` given that misbehaviour has been detected and verified.
```

> - `UpdateState` : updates and stores as necessary any associated information for an IBC client, such as the `ClientState` and corresponding `ConsensusState`. An error is returned if `ClientMessage` is of type `Misbehaviour`.

Reviewer: ```suggestion
- `UpdateState` updates and stores as necessary any associated information for an IBC client, such as the `ClientState` and corresponding `ConsensusState`. An error is returned if `ClientMessage` is of type `Misbehaviour`.
```

Reviewer: @catShaark can you commit this suggestion?

The `CheckMisbehaviourAndUpdateState` function has been removed from `ClientState` interface. This functionality is now encapsulated by the usage of `VerifyClientMessage`, `CheckForMisbehaviour`, `UpdateStateOnMisbehaviour`, `UpdateState`.

The function `GetTimestampAtHeight` has been added to the `ClientState` interface. It should return the timestamp for a consensus state associated with the provided height.

> ### Header and Misbehaviour

Reviewer: ```suggestion
### `Header` and `Misbehaviour`
```

> `exported.Header` and `exported.Misbehaviour` interface types have been merged and renamed to `ClientMessage` interface.

Author: yess

> `exported.Header` and `exported.Misbehaviour` interface types have been merged and renamed to `ClientMessage` interface.

Author: should I add that we change the verify function of client state ?

> ### ConsensusState

Reviewer: ```suggestion
### `ConsensusState`
```

> The `GetRoot` function has been removed from consensus state interface.

Reviewer: ```suggestion
The `GetRoot` function has been removed from consensus state interface since it was not used by core IBC.
```

### Light client implementations

09-localhost light client implementation has been removed because it is currently non-functional.

### Client Keeper

> Keeper function `CheckMisbehaviourAndUpdateState` has been removed since function `UpdateClient` can now handle updating `ClientState` on `ClientMessage` type which can be any `Misbehaviour` implementations.  

Reviewer: ```suggestion
Keeper function `CheckMisbehaviourAndUpdateState` has been removed since function `UpdateClient` can now handle updating `ClientState` on `ClientMessage` type which can be any `Misbehaviour` implementation.  
```

### SDK Message

> `MsgSubmitMisbehaviour` is deprecated since `MsgUpdateClient` can now submit a `ClientMessage` type which can be any `Misbehaviour` implementations.

Reviewer: ```suggestion
`MsgSubmitMisbehaviour` is deprecated since `MsgUpdateClient` can now submit a `ClientMessage` type which can be any `Misbehaviour` implementation.
```

Author: But we have to use plural after any.

Reviewer: This is really a nit, so I think either it's fine.

## IBC Light Clients

> ### Light client and ClientState

Reviewer 2: ```suggestion
### ClientState interface changes
```

The `VerifyUpgradeAndUpdateState` function has been modified. The client state and consensus state return value has been removed. 

> Light clients **must** set the updated client state and consensus state in the client store after verifying a valid client upgrade.

Reviewer 2: ```suggestion
Light clients **must** handle all management of client and consensus states including the setting of updated client state and consensus state in the client store.
```

> The `CheckHeaderAndUpdateState` function has been split into 4 new functions: `VerifyClientMessage`, `CheckForMisbehaviour`, `UpdateState`, 
> `UpdateStateOnMisbehaviour`

Reviewer 2: ```suggestion
The `CheckHeaderAndUpdateState` function has been split into 4 new functions: 
- `VerifyClientMessage` 
- `CheckForMisbehaviour` 
- `UpdateStateOnMisbehaviour`
- `UpdateState`, 

```

We should add a short description for each function. Maybe you could use the godoc from #1281?

Author: yesss

> Light client implementations now need to manage setting of client and consensus states for these interface functions `UpdateState`, `UpdateStateOnMisbehaviour`, `VerifyUpgradeAndUpdateState`, `CheckSubstituteAndUpdateState`
> 

Reviewer 2: ```suggestion
```

I think we can remove this and state the light client must handle all management of client and consensus states, including setting updated client/consensus states

> The `CheckMisbehaviourAndUpdateState` function has been removed from `ClientState` interface

Reviewer 2: ```suggestion
The `CheckMisbehaviourAndUpdateState` function has been removed from `ClientState` interface. This functionality is now encapsulated by the usage of `VerifyClientMessage`, `CheckForMisbehaviour`, `UpdateStateOnMisbehaviour`, `UpdateState`.

```

> The `GetTimestampAtHeight` has been added to the `ClientState` interface

Reviewer 2: ```suggestion
The function `GetTimestampAtHeight` has been added to the `ClientState` interface. It should return the timestamp for a consensus state associated with the provided height.
```

### Header and Misbehaviour

> `exported.Header` and `exported.Misbehaviour` interface types have been merged and renamed to `ClientMessage` interface

Reviewer 2: ```suggestion
`exported.Header` and `exported.Misbehaviour` interface types have been merged and renamed to `ClientMessage`. interface
```

### ConsensusState

> The `GetRoot` function has been removed from consensus state interface

Reviewer 2: ```suggestion
The `GetRoot` function has been removed from consensus state interface.
```

### `ClientState` interface changes

> The `VerifyUpgradeAndUpdateState` function has been modified. The client state and consensus state return values has been removed.

Author: ohh jeez, sorry I forgot this.

> The `VerifyUpgradeAndUpdateState` function has been modified. The client state and consensus state return values has been removed.

Reviewer 3: just a nit :) great work!
